### ISSUE TITLE
As a user, I want to query Toggl Reports in a codeblock to use as part of daily/weekly/... review notes.

### ISSUE THREAD

**Issue by mcndt** [mcndt/obsidian-toggl-integration#27]
Before I make a feature request, I wanted to make sure I understood this roadmap item. If implemented, would this enable a user to display, for instance, all the Toggl entries for a specific date in an Obisidian note? If so, I'd strongly +1 this feature.

I would love to include a Toggl code block in my daily and weekly review notes. Thanks!
Jack

_Originally posted by @jebeaudoin in https://github.com/mcndt/obsidian-toggl-integration/discussions/26_

**Comment by cor** [comment]
I would love this functionality!

**Comment by tonyxiao** [comment]
That would be so sick!

**Comment by shabegom** [comment]
How dare you call me a "user"! But yeah would love this or some way to log Toggl times to my DNP.

**Comment by mcndt** [comment]
@tonyxiao @cor  @shabegom 

Hey everyone, as I am starting my final graduate year I am eager to develop this feature because I myself would get a lot of use out of it. I have just whipped up a first draft of the query language for creating these reports and I would like some feedback from y'all to see if this covers all your needs 😄 

What I think is still lacking currently, but that I am not sure is useful is: 

- including/excluding by client instead of project
- Querying different workspaces than the one set in the plugin settings

These reports are also heavily drawn from Toggl's own report pages. If you have any more creative or different ways you would like to visualize your time entries please let me know.

### Summary reports

very similar to summary report on track.toggl.com

```js
// report type = summary
SUMMARY 

// Add one of these as time interval selector
WEEKLY                          // get summary for this week
MONTHLY                         // get summary for this month
FROM {date0} TO {date1}         // get summary for date range
PREVIOUS {x days, x weeks, ...} // get summary for previous dates

// Optional, add one of these to narrow query
[INCLUDE 'project A', 'project B', ...] // whitelist projects
[EXCLUDE 'project A', 'project B', ...] // blacklist projects

// Optional, add any of these to limit visualizations
[HIDE PIE]   // hides colored area plot 
[HIDE BAR]   // hides time bars (e.g. bar plot with bars for each day of week)
[HIDE LIST]  // hides list of projects with totals
```

For example, to get an overview of the last week, do not include the project “hobby project” without the pie plot:

```js
SUMMARY WEEKLY
EXCLUDE 'hobby project'
HIDE PIE
```

### Detailed reports

Similar to track.toggl.com/timer list view, but can also be grouped by project.

```js
// report type = time entry list
LIST

// Add one of these as time interval selector.
WEEKLY                          // get summary for this week
MONTHLY                         // get summary for this month
FROM {date0} TO {date1}         // get summary for date range
PREVIOUS {x days, x weeks, ...} // get summary for previous dates

// Optional, add one of these to narrow query
[INCLUDE 'project A', 'project B', ...] // whitelist projects
[EXCLUDE 'project A', 'project B', ...] // blacklist projects

// Optional, group the time entries by project or day. 
// This will show a header for each group with some aggregate stats,
// e.g. total duration per group
[GROUP BY {(DATE|PROJECT)}]

// Optional, change sort order by date
[SORT {ASC|DESC}]

// Optional, limit the amount of results
[LIMIT {N}] // Limit the amount of results
```

Example: give me all time entries from the past 7 days for the project “Obsidian Plugin Development”:

```js
LIST PREVIOUS "7 days"
INCLUDE "Obsidian Plugin Development"
GROUP BY DATE
```

**Comment by shabegom** [comment]
I think this is looks excellent!

- I only have one workspace so wouldn't use multiple workspaces
- I have multiple clients but they break down into specific projects for each client. I would prefer to do project based inclusion. 
- Im most likely going to want a daily report specific to a date. Nice to have would be an easy way to supply one date instead of a range. Range is simple enough though, so not an essential addition.

```
LIST FROM <% tp.file.title %> TO <% tp.file.title %>
INCLUDE "Super Secret Project"
```

**Comment by mcndt** [comment]
@shabegom That's an excellent use case that I forgot to include. I'll make sure it is minimal effort to get a day report, I'm not a fan of having to do a range of one date either!

**Comment by tonyxiao** [comment]
I would love to be able to filter by 'user', I'm in a Toggl workspace with multiple users (including people I manage) and I'd love to be able to both see my time as well as theirs.

**Comment by mcndt** [comment]
@tonyxiao I will keep that in mind! For now my focus will be developing the query language as presented here (a bit of a technical learning opportunity by itself), but when that is finished I can consider adding support for multiple users.

**Comment by mcndt** [comment]
Hi everyone. There have been quite a few months of inactivity on this project as I have been preoccupied with my Master dissertation. However, with the holidays I have been able to get a lot of work done on this feature.

If all goes well, I think I can release this feature set by the end of January.

Here's a preview:

![image](https://user-images.githubusercontent.com/23149353/147885485-5cb0fc8b-b24e-4764-ad93-1ed4a7829a44.png)

renders into:

![image](https://user-images.githubusercontent.com/23149353/147885493-674a7037-788d-4288-ae5c-000f26ae40eb.png)

**Comment by jebeaudoin** [comment]
C’est formidable!

**Comment by mcndt** [comment]
This feature is now available in 0.4.0! Don't hesitate to shoot me feedback regarding bugs or use cases on the release discussion page: 

https://github.com/mcndt/obsidian-toggl-integration/discussions/38

**Comment by BrianBatchelder** [comment]
I think you need to update your [Development Roadmap](https://github.com/mcndt/obsidian-toggl-integration/projects/1). This (and one other issue) are shown as "Next Release" when, in fact, I believe you have already released them.
